// @diplodoc/transform/lib/yfmlint/yfmlint.js
"use strict";
const log_1 = require("../log");
const index = {
    // Default state for all rules
    default: true,
    'log-levels': {
        MD001: log_1.LogLevels.DISABLED,
        MD002: log_1.LogLevels.DISABLED,
        MD003: log_1.LogLevels.DISABLED,
        MD004: log_1.LogLevels.DISABLED,
        MD005: log_1.LogLevels.DISABLED,
        MD006: log_1.LogLevels.DISABLED,
        MD007: log_1.LogLevels.DISABLED,
        MD009: log_1.LogLevels.DISABLED,
        MD010: log_1.LogLevels.DISABLED,
        MD011: log_1.LogLevels.DISABLED,
        MD012: log_1.LogLevels.DISABLED,
        MD013: log_1.LogLevels.DISABLED,
        MD014: log_1.LogLevels.DISABLED,
        MD018: log_1.LogLevels.DISABLED,
        MD019: log_1.LogLevels.DISABLED,
        MD020: log_1.LogLevels.DISABLED,
        MD021: log_1.LogLevels.DISABLED,
        MD022: log_1.LogLevels.DISABLED,
        MD023: log_1.LogLevels.DISABLED,
        MD024: log_1.LogLevels.DISABLED,
        MD025: log_1.LogLevels.DISABLED,
        MD026: log_1.LogLevels.DISABLED,
        MD027: log_1.LogLevels.DISABLED,
        MD028: log_1.LogLevels.DISABLED,
        MD029: log_1.LogLevels.DISABLED,
        MD030: log_1.LogLevels.DISABLED,
        MD031: log_1.LogLevels.DISABLED,
        MD032: log_1.LogLevels.DISABLED,
        MD033: log_1.LogLevels.DISABLED,
        MD034: log_1.LogLevels.DISABLED,
        MD035: log_1.LogLevels.DISABLED,
        MD036: log_1.LogLevels.DISABLED,
        MD037: log_1.LogLevels.DISABLED,
        MD038: log_1.LogLevels.DISABLED,
        MD039: log_1.LogLevels.DISABLED,
        MD040: log_1.LogLevels.DISABLED,
        MD041: log_1.LogLevels.DISABLED,
        MD042: log_1.LogLevels.DISABLED,
        MD043: log_1.LogLevels.DISABLED,
        MD044: log_1.LogLevels.DISABLED,
        MD045: log_1.LogLevels.DISABLED,
        MD046: log_1.LogLevels.DISABLED,
        MD047: log_1.LogLevels.DISABLED,
        MD048: log_1.LogLevels.DISABLED,
        MD049: log_1.LogLevels.DISABLED,
        MD050: log_1.LogLevels.DISABLED,
        MD051: log_1.LogLevels.DISABLED,
        MD052: log_1.LogLevels.DISABLED,
        MD053: log_1.LogLevels.DISABLED,
        MD054: log_1.LogLevels.DISABLED,
        MD055: log_1.LogLevels.DISABLED,
        MD056: log_1.LogLevels.DISABLED,
        YFM001: log_1.LogLevels.WARN,
        YFM002: log_1.LogLevels.WARN,
        YFM003: log_1.LogLevels.ERROR,
        YFM004: log_1.LogLevels.ERROR,
        YFM005: log_1.LogLevels.ERROR,
        YFM006: log_1.LogLevels.WARN,
        YFM007: log_1.LogLevels.WARN,
        YFM008: log_1.LogLevels.WARN,
        YFM009: log_1.LogLevels.ERROR, // Term definition used not at the end of file
    },
    // Inline code length
    YFM001: {
        maximum: 100,
    },
};
module.exports = index;
