// @diplodoc/transform/lib/plugins/file/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileClassName = exports.PREFIX_LENGTH = exports.PREFIX = exports.FILE_TOKEN = exports.REQUIRED_ATTRS = exports.KNOWN_ATTRS = exports.RULE_NAME = exports.FILE_TO_LINK_ATTRS_MAP = exports.LinkHtmlAttr = exports.FileSpecialAttr = void 0;
var FileSpecialAttr;
(function (FileSpecialAttr) {
    FileSpecialAttr["Src"] = "src";
    FileSpecialAttr["Name"] = "name";
    FileSpecialAttr["Lang"] = "lang";
})(FileSpecialAttr = exports.FileSpecialAttr || (exports.FileSpecialAttr = {}));
// https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attributes
var LinkHtmlAttr;
(function (LinkHtmlAttr) {
    LinkHtmlAttr["Download"] = "download";
    LinkHtmlAttr["Href"] = "href";
    LinkHtmlAttr["HrefLang"] = "hreflang";
    LinkHtmlAttr["Media"] = "media";
    LinkHtmlAttr["Ping"] = "ping";
    LinkHtmlAttr["ReferrerPolicy"] = "referrerpolicy";
    LinkHtmlAttr["Rel"] = "rel";
    LinkHtmlAttr["Target"] = "target";
    LinkHtmlAttr["Type"] = "type";
})(LinkHtmlAttr = exports.LinkHtmlAttr || (exports.LinkHtmlAttr = {}));
exports.FILE_TO_LINK_ATTRS_MAP = {
    [FileSpecialAttr.Src]: LinkHtmlAttr.Href,
    [FileSpecialAttr.Name]: LinkHtmlAttr.Download,
    [FileSpecialAttr.Lang]: LinkHtmlAttr.HrefLang,
};
exports.RULE_NAME = 'yfm_file_inline';
exports.KNOWN_ATTRS = [
    FileSpecialAttr.Src,
    FileSpecialAttr.Name,
    FileSpecialAttr.Lang,
    LinkHtmlAttr.ReferrerPolicy,
    LinkHtmlAttr.Rel,
    LinkHtmlAttr.Target,
    LinkHtmlAttr.Type,
];
exports.REQUIRED_ATTRS = [FileSpecialAttr.Src, FileSpecialAttr.Name];
exports.FILE_TOKEN = 'yfm_file';
exports.PREFIX = '{% file ';
exports.PREFIX_LENGTH = exports.PREFIX.length;
var FileClassName;
(function (FileClassName) {
    FileClassName["Link"] = "yfm-file";
    FileClassName["Icon"] = "yfm-file__icon";
})(FileClassName = exports.FileClassName || (exports.FileClassName = {}));
